### DISCUSSION TITLE
Create a 512 by 256 bitmap in Digital?

### DISCUSSION THREAD

**Discussion by VerdichteteEnergie** [hneemann/Digital#1117] [Q&A] [answered]
I created an 8K RAM bar.
This should serve as graphics memory and save and output in 16-bit.
I would like to output the bits in 512 by 256 pixels on a monochrome screen (black and white). Each bit should stand for one pixel.
The pixels of the bits should be arranged next to each other from top left to right.
For example, if the first register of RAM stores 1111 1111 1111 1111, then the first 16 pixels in the first row should be black from left to right.

My goal is to create a similar screen as in the nand2tetris course.

![Screenshot 2023-02-14 071848](https://user-images.githubusercontent.com/76449084/218655299-ad93ec78-1849-452d-861f-8a461a9dc3e2.png)
_The text box is just a placeholder for the screen._

Is it possible to create such a screen in digital?

**Comment by hneemann** [comment]
You have to implement such a component as a java-plugin.
But Ivan Toshkov has already implemented such a display.
See here: https://github.com/itoshkov/digital-n2t

**Comment by VerdichteteEnergie** [comment, marked as answer]
Many thanks for your response!

I downloaded and unpacked the raw data from the library you linked and replaced the jar file.
Then I created the library using maven.
![Screenshot 2023-02-14 083012](https://user-images.githubusercontent.com/76449084/218671468-4422356d-7523-4a95-9c5a-60863eed9df0.png)
A new folder has been created in the directory named target. In this folder is a Java file called "digitaln2t-1.0-SNAPSHOT.jar".
I added this in Digital under "Settings" > "Java Library".
![Screenshot 2023-02-14 083822](https://user-images.githubusercontent.com/76449084/218671549-9aa9fce8-c211-4a0f-ac66-2af44f4efe27.png)
After that, I was presented with a new tab called "nand2tetris". When I open the tab, I see a RAM chip "Hack Display" in the menu.
![Screenshot 2023-02-14 083204](https://user-images.githubusercontent.com/76449084/218671613-71c81f74-cbfc-4cac-a4e5-2f48ef955bc5.png)
But even when I start the simulation, I don't see a screen.
![Screenshot 2023-02-14 083226](https://user-images.githubusercontent.com/76449084/218671642-cc35b267-78cf-4854-816f-de25895269ac.png)
Do I have to connect something to the data output of the "Hack-Display" chip?

I also noticed that in the simulation the real time clock stops as soon as I enable the str input of the hack display chip.

**Reply by hneemann** [reply]
You have to connect the D-output only if you want to read from the RAM component.

On my machine it works just fine. 
I have used my own fork to test (https://github.com/hneemann/digital-n2t) but the code seems to be the same.
Are there some error logs in your home folder?

**Reply by VerdichteteEnergie** [reply]
I also tried your new link. Unfortunately, there is still no pop-up with a screen.

I think that I don't see any error logs. Here is what my folder looks like:

My home folder for digital is C:\Users\User\OneDrive\Data\Digital - NAND_2_TETRIS
![Screenshot 2023-02-14 105704](https://user-images.githubusercontent.com/76449084/218702458-74997781-0195-47bd-922e-ddb929e4eece.png)
The folder with the new library and copied digital.jar is C:\Users\User\OneDrive\Desktop\digital-n2t\digital-n2t-master
![Screenshot 2023-02-14 105650](https://user-images.githubusercontent.com/76449084/218702484-eb8f69c4-2952-414b-90bf-fe989110a4ef.png)

**Reply by hneemann** [reply]
Error logs are created in your home folder. 
I'm not sure where the home folder is located on a Win10 machine though. I have been a Linux only user for too many years. :smile:

**Reply by VerdichteteEnergie** [reply]
I'm sorry.

In the Windows 11 "Event Viewer" I looked at every error message and warnings today.

I couldn't find anything related to the digital app.
![Screenshot 2023-02-14 114349](https://user-images.githubusercontent.com/76449084/218713732-3d79a19e-ca35-47bf-805e-6ad5aac24b6c.png)

**Reply by VerdichteteEnergie** [reply]
Hmm, I don't get any output from the chip either when I've stored something into the hackdisplay chip before.

As I said, the clock doesn't seem to work properly with the chip for me. Real time clock not working. Manual switching causes graphic delays.

Do you think there is a way to realize the bitmap screen with 8K RAM and LED matrix?

**Reply by hneemann** [reply]
The files look like this: **Digital_2022-08-09_11-58-12.log**

Such a large display cannot be realized with the LED matrix. You could use the built-in graphic display, but this has a different memory layout. Of course this doesn't matter if you write the code yourself, but programs written for the n2t display won't run then. Maybe it is possible to build a converter circuit that can be written like an n2t display, and uses the built-in graphic display for the output. I have to think about that.

**Reply by VerdichteteEnergie** [reply]
Thank you! Now I found the logs. 

As I can see there were a lot of log files today. But all have the same content. Unfortunately I don't understand it. Here is the upload of the last:
[Digital_2023-02-14_12-02-59.log](https://github.com/hneemann/Digital/files/10731923/Digital_2023-02-14_12-02-59.log)

**Reply by hneemann** [reply]
Please try this jar file: [n2t-display.zip](https://github.com/hneemann/Digital/files/10732209/n2t-display.zip)

**Reply by VerdichteteEnergie** [reply]
It works!

Thank you so much!

You're a hero!
